iothub_client/src/iothubtransport_mqtt_common.c

```diff
@@ -1291,329 +1338,409 @@ static bool isSystemProperty(const char* tokenData)
+        // Substract one from queue length to reflect this having an extra # we subscribed to
+        propertiesStart = topic_name + (strlen(input_queue_topic) - 1);
+    }
+    else if (type == IOTHUB_TYPE_TELEMETRY)
+    {
+        const char* mqtt_message_queue_topic = STRING_c_str(transportData->topic_MqttMessage);
+        // Substract one from queue length to reflect this having an extra # we subscribed to
+        propertiesStart = topic_name + (strlen(mqtt_message_queue_topic) - 1);
+    }
+    else
+    {
+        LogError("ERROR: type %d is not expected", type);
+        propertiesStart = NULL;
+    }
+
+    return propertiesStart;
+}
+
+//
+// AddApplicationProperty adds the custom key/value property name from the incoming MQTT PUBLISH to the iotHubMessage
+// we will ultimately deliver to the application on its callback.
+//
+static int AddApplicationPropertyToMessage(MAP_HANDLE propertyMap, const char* propertyNameAndValue, size_t propertyNameLength, const char* propertyValue, bool auto_url_encode_decode)
+{
+    int result = 0;
+
+    char* propertyNameCopy = NULL;
+
+    // We need to make a copy of propertyName at this point; we don't own the buffer it's part of
+    // so we can't just sneak a temporary '\0' in there.  We don't need to make a copy of propertyValue
+    // since its part of an otherwise null terminated string.
+    if ((propertyNameCopy = (char*)malloc(propertyNameLength + 1)) == NULL)
+    {
+        LogError("Failed allocating property information");
+        result = MU_FAILURE;
+    }
+    else
     {
-        MAP_HANDLE propertyMap = IoTHubMessage_Properties(IoTHubMessage);
-        if (propertyMap == NULL)
+        memcpy(propertyNameCopy, propertyNameAndValue, propertyNameLength);
+        propertyNameCopy[propertyNameLength] = '\0';
+    
+        if (auto_url_encode_decode)
         {
-            LogError("Failure to retrieve IoTHubMessage_properties.");
+            STRING_HANDLE propName_decoded = URL_DecodeString(propertyNameCopy);
+            STRING_HANDLE propValue_decoded = URL_DecodeString(propertyValue);
+            if (propName_decoded == NULL || propValue_decoded == NULL)
+            {
+                LogError("Failed to URL decode property");
+                result = MU_FAILURE;
+            }
+            else if (Map_AddOrUpdate(propertyMap, STRING_c_str(propName_decoded), STRING_c_str(propValue_decoded)) != MAP_OK)
+            {
+                LogError("Map_AddOrUpdate failed.");
+                result = MU_FAILURE;
+            }
+            else
+            {
+                result = 0;
+            }
+            STRING_delete(propValue_decoded);
+            STRING_delete(propName_decoded);
+        }
+        else if (Map_AddOrUpdate(propertyMap, propertyNameCopy, propertyValue) != MAP_OK)
+        {
+            LogError("Map_AddOrUpdate failed.");
+            result = MU_FAILURE;
+        }
+    }
+
+    free(propertyNameCopy);
```

Comment:
This should be after line 1605, in the block after malloc succeeds. Otherwise it's called in case the variable is still null.